tools/satsolver-main.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/CDCL.h>
+#include <vector>
+#include <string>
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <sstream>
+
+using namespace solidity::util;
+using std::vector;
+using std::string;
+using std::cout;
+using std::endl;
+using std::optional;
+using std::nullopt;
+using std::make_pair;
+
+const int verbose = 0;
+vector<string> cutStringBySpace(string& input);
+optional<vector<Literal>> parseLine(std::string& line);
+std::pair<vector<vector<Literal>>, size_t> readCNFFile(const string& fname);
+size_t getNumVars(const vector<vector<Literal>>& cls);
+
+
+vector<string> cutStringBySpace(string& input)
+{
+    vector<string> parts;
+
+    size_t pos = 0;
+    while ((pos = input.find(" ")) != string::npos)
+	{
+        parts.push_back(input.substr(0, pos));
+        input.erase(0, pos + 1);
+    }
+    parts.push_back(input);
+    return parts;
+}
+
+optional<vector<Literal>> parseLine(std::string& line)
+{
+	vector<Literal> cl;
+	bool end_of_clause = false;
+	auto parts = cutStringBySpace(line);
+	for (const auto& part: parts)
+	{
+		assert(!end_of_clause);
+		const long lit = std::stol(part);
+		const long var = std::abs(lit);
+		if (var == 0)
+		{
+			//end of clause
+			end_of_clause = true;
+			continue;
```

Comment:
Shouldn't this be a `break`?

Reply:
I did this so we catch if someone made a mistake, like this in the CNF input:
```
1 2 3 0 4
```

I set `end_of_clause` once I reach `0` and it `continue`-s. Then the 1st line after the `for` is `assert(!end_of_clause)`. This catches the issue:

```
$ cat a.cnf 
p cnf 2 4
1 2 0 3
-1 -2 0
1 -2 0
-1 2 0
$ ./satsolver a.cnf proof.frat 
satsolver: /solidity/tools/satsolver-main.cpp:64: std::optional<std::vector<solidity::util::Literal> > parseLine(std::string&): Assertion `!end_of_clause' failed.
Aborted (core dumped)
```

However, catching it with an `assert` is probably wrong. I now emit a written error and `exit(-1)`, which is better. It's not a code bug (i.e. not a place for an `assert`), but an user bug:

```
$ ./satsolver a.cnf proof.frat 
ERROR: trailing elements after finishing `0` at the end of a clause in CNF file
```

In my experience, one can chase their tail for a long time on issues like this, where their CNF generation script is incorrect. Simple checks like this tend to help the user (often, me :laughing: )

Let me know what you think!

Reply:
Ah ok I see! The more relaxed version ( https://www.cs.utexas.edu/users/moore/acl2/manuals/current/manual/index-seo.php/SATLINK____DIMACS ) does not require the line to end with a zero, though. Not sure if this is important to us.